<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Общие-данные" data-toc-modified-id="Общие-данные-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Общие данные</a></span><ul class="toc-item"><li><span><a href="#Цель-исследования" data-toc-modified-id="Цель-исследования-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Цель исследования</a></span></li><li><span><a href="#План-работы" data-toc-modified-id="План-работы-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>План работы</a></span></li></ul></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Описание данных</a></span></li><li><span><a href="#Исследование-таблиц:" data-toc-modified-id="Исследование-таблиц:-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Исследование таблиц:</a></span><ul class="toc-item"><li><span><a href="#Вывод-первых-строк:" data-toc-modified-id="Вывод-первых-строк:-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Вывод первых строк:</a></span></li><li><span><a href="#Вывод-по-исследованию-таблиц" data-toc-modified-id="Вывод-по-исследованию-таблиц-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Вывод по исследованию таблиц</a></span></li></ul></li><li><span><a href="#Решение-заданий-запросами-SQL:" data-toc-modified-id="Решение-заданий-запросами-SQL:-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Решение заданий запросами SQL:</a></span><ul class="toc-item"><li><span><a href="#Подсчет-количества-книг,-вышедших-после-01.01.2000-г." data-toc-modified-id="Подсчет-количества-книг,-вышедших-после-01.01.2000-г.-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Подсчет количества книг, вышедших после 01.01.2000 г.</a></span><ul class="toc-item"><li><span><a href="#Вывод-по-задаче" data-toc-modified-id="Вывод-по-задаче-4.1.1"><span class="toc-item-num">4.1.1&nbsp;&nbsp;</span>Вывод по задаче</a></span></li></ul></li><li><span><a href="#Для-каждой-книги-посчитать-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-посчитать-количество-обзоров-и-среднюю-оценку-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Для каждой книги посчитать количество обзоров и среднюю оценку</a></span><ul class="toc-item"><li><span><a href="#Вывод-по-задаче" data-toc-modified-id="Вывод-по-задаче-4.2.1"><span class="toc-item-num">4.2.1&nbsp;&nbsp;</span>Вывод по задаче</a></span></li></ul></li><li><span><a href="#Определение-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц---таким-образом,-исключим-брошюры-из-анализа" data-toc-modified-id="Определение-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц---таким-образом,-исключим-брошюры-из-анализа-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Определение издательство, которое выпустило наибольшее число книг толще 50 страниц - таким образом, исключим брошюры из анализа</a></span><ul class="toc-item"><li><span><a href="#Вывод-по-задаче" data-toc-modified-id="Вывод-по-задаче-4.3.1"><span class="toc-item-num">4.3.1&nbsp;&nbsp;</span>Вывод по задаче</a></span></li></ul></li><li><span><a href="#Определение-автора-с-самой-высокой-средней-оценкой-книг---необходимо-учитывать-только-книги-с-50-и-более-оценками" data-toc-modified-id="Определение-автора-с-самой-высокой-средней-оценкой-книг---необходимо-учитывать-только-книги-с-50-и-более-оценками-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Определение автора с самой высокой средней оценкой книг - необходимо учитывать только книги с 50 и более оценками</a></span><ul class="toc-item"><li><span><a href="#Вывод-по-задаче" data-toc-modified-id="Вывод-по-задаче-4.4.1"><span class="toc-item-num">4.4.1&nbsp;&nbsp;</span>Вывод по задаче</a></span></li></ul></li><li><span><a href="#Подсчет-среднего-количества-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="Подсчет-среднего-количества-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Подсчет среднего количества обзоров от пользователей, которые поставили больше 50 оценок</a></span><ul class="toc-item"><li><span><a href="#Вывод-по-задаче" data-toc-modified-id="Вывод-по-задаче-4.5.1"><span class="toc-item-num">4.5.1&nbsp;&nbsp;</span>Вывод по задаче</a></span></li></ul></li></ul></li></ul></div>

# Анализ базы данных

## Общие данные

Наша компания приобрела крупный сервис для чтения книг по подписке. Необходимо проанализировать базу данных. В ней содержится информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. На основе этих данных можно будет сформулировать ценностное предложение для нового продукта.

### Цель исследования

**Цель исследования** - анализ базы данных сервиса чтения книг по подписке

### План работы

**2. Описание данных**

**3. Исследование таблиц:**
   - вывод первых строк
   - вывод по исследованию таблиц
   
**4. Решение заданий запросами SQL:**
   - подсчет количества книг, вышедших после 01.01.2000 г.
      - вывод по задаче
   - для каждой книги посчитать количество обзоров и среднюю оценку
      - вывод по задаче
   - определение издательство, которое выпустило наибольшее число книг толще 50 страниц - таким образом, исключим брошюры из анализа
      - вывод по задаче
   - определение автора с самой высокой средней оценкой книг - необходимо учитывать только книги с 50 и более оценками
      - вывод по задаче
   - подсчет среднего количества обзоров от пользователей, которые поставили больше 50 оценок
      - вывод по задаче
   
**5. Общий вывод:**

## Описание данных

Согласно прилагаемой документации, в базе данных 4 таблицы со следующим содержимым:

1. Таблица `books` содержит данные о книгах.

**Структура books:**

- **`book_id`** — идентификатор книги
- **`author_id`** — идентификатор автора
- **`title`** — название книги
- **`num_pages`** — количество страниц
- **`publication_date`** — лата публикации книги
- **`publisher_id`** — идентификатор издателя

2. Таблица `autors` содержит данные об авторах.

**Структура autors:**

- **`author_id`** — идентификатор автора
- **`author`** — имя автора

3. Таблица `publishers` содержит данные об издательствах.

**Структура publishers:**

- **`publisher_id`** — идентификатор издательства
- **`publisher`** — название издательства

4. Таблица `ratings` содержит данные о пользовательских оценках книг.

**Структура ratings:**

- **`rating_id`** — идентификатор оценки
- **`book_id`** — идентификатор книги
- **`username`** — имя пользователя, оставившего оценку
- **`rating`** — оценка книги

5. Таблица `reviews` содержит данные о пользовательских обзорах на книги.

**Структура reviews:**

- **`review_id`** — идентификатор обзора
- **`book_id`** — идентификатор книги
- **`username`** — имя пользователя, написавшего обзор
- **`text`** — текст обзора

## Исследование таблиц:

Подгружаем библиотеки, которые понадобятся для нашего исследования:

In [1]:
import pandas as pd
from sqlalchemy import create_engine

Устанавливаем параметры:

In [2]:
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

### Вывод первых строк:

Таблица **books** - первые строки:

In [3]:
query = '''
        SELECT *
        FROM books
        '''
books = pd.io.sql.read_sql(query, con = engine)
print(books.info())
books.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
book_id             1000 non-null int64
author_id           1000 non-null int64
title               1000 non-null object
num_pages           1000 non-null int64
publication_date    1000 non-null object
publisher_id        1000 non-null int64
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
None


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Анализ таблицы `books`:
- пропусков нет
- в библиоткеке 1000 книг
- дата не соответствует типу данных для дат

Таблица **authors** - первые строки:

In [4]:
query = '''
        SELECT *
        FROM authors
        '''
authors = pd.io.sql.read_sql(query, con = engine)
print(authors.info())
authors.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
author_id    636 non-null int64
author       636 non-null object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
None


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

Анализ таблицы `authors`:
- пропусков нет
- в базе данных 636 записей об авторах

Таблица **publishers** - первые строки:

In [5]:
query = '''
        SELECT *
        FROM publishers
        '''
publishers = pd.io.sql.read_sql(query, con = engine)
print(publishers.info())
publishers.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
publisher_id    340 non-null int64
publisher       340 non-null object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
None


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

Анализ таблицы `publishers`:
- пропусков нет
- в базе данных 340 записей об издательствах

Таблица **ratings** - первые строки:

In [6]:
query = '''
        SELECT *
        FROM ratings
        '''
ratings = pd.io.sql.read_sql(query, con = engine)
print(ratings.info())
ratings.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
rating_id    6456 non-null int64
book_id      6456 non-null int64
username     6456 non-null object
rating       6456 non-null int64
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
None


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

Анализ таблицы `ratings`:
- пропусков нет
- в базе данных 6456 записей об оценках пользователей

Таблица **reviews** - первые строки:

In [7]:
query = '''
        SELECT *
        FROM reviews
        '''
reviews = pd.io.sql.read_sql(query, con = engine)
print(reviews.info())
reviews.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
review_id    2793 non-null int64
book_id      2793 non-null int64
username     2793 non-null object
text         2793 non-null object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
None


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Анализ таблицы `reviews`:
- пропусков нет
- в базе данных 2793 записей-ревью от пользователей

### Вывод по исследованию таблиц

По всем таблицам получены следующие сведения:
- пропуски отсутствуют во всех таблицах
- дата публикации в таблице `books` не соответствует типу данных для даты

## Решение заданий запросами SQL:

### Подсчет количества книг, вышедших после 01.01.2000 г.

In [8]:
query = '''
        SELECT COUNT(book_id)
        FROM books
        WHERE publication_date >= '2000-01-01'
        '''

pd.io.sql.read_sql(query, con = engine)

count
0    821

#### Вывод по задаче

- Количество книг, выпущенных после 01.01.2000 - **821 шт.**
- Анализ даты показывает, что тип данных в базе данных указан верно. Pandas(Python) - распознает как текст.

### Для каждой книги посчитать количество обзоров и среднюю оценку

In [9]:
query = '''
        WITH
        rat AS (SELECT book_id,
                       AVG(rating) AS mean_rat
                FROM ratings
                GROUP BY book_id),
        rev AS (SELECT book_id,
                       COUNT(*) AS count_rev
                FROM reviews
                GROUP BY book_id)

        SELECT b.book_id,
               rat.mean_rat,
               rev.count_rev
        FROM books AS b
        LEFT JOIN rat ON b.book_id=rat.book_id
        LEFT JOIN rev ON b.book_id=rev.book_id
        ORDER BY mean_rat DESC, count_rev DESC
        '''

pd.io.sql.read_sql(query, con = engine)

book_id  mean_rat  count_rev
0        672      5.00        NaN
1         17      5.00        4.0
2        553      5.00        3.0
3        444      5.00        3.0
4        516      5.00        2.0
..       ...       ...        ...
995      915      2.25        3.0
996      202      2.00        3.0
997      371      2.00        2.0
998      316      2.00        2.0
999      303      1.50        2.0

[1000 rows x 3 columns]

#### Вывод по задаче

- **по некоторым книгам нет ревью от пользователей**

### Определение издательство, которое выпустило наибольшее число книг толще 50 страниц - таким образом, исключим брошюры из анализа

In [10]:
query = '''
        SELECT pub.publisher_id, 
               pub.publisher,
               COUNT (book_id) AS count_books
        FROM books AS b
        LEFT JOIN publishers AS pub ON b.publisher_id=pub.publisher_id
        WHERE num_pages > 50
        GROUP BY pub.publisher_id, pub.publisher
        ORDER BY count_books DESC
        LIMIT 5
        '''

pd.io.sql.read_sql(query, con = engine)

publisher_id                 publisher  count_books
0           212             Penguin Books           42
1           309                   Vintage           31
2           116  Grand Central Publishing           25
3           217          Penguin Classics           24
4            35                    Bantam           19

#### Вывод по задаче

- **издательство** с самым **большим количеством изданных книг** (больше 50 страниц) - **Penguin Books: 42 шт**

### Определение автора с самой высокой средней оценкой книг - необходимо учитывать только книги с 50 и более оценками

In [11]:
query = '''
        WITH
        rat AS (SELECT book_id,
                       AVG(rating) AS mean_rat
                FROM ratings
                GROUP BY book_id
                HAVING COUNT(*) > 50)

        SELECT a.author,
               ROUND(rat.mean_rat, 2)
        FROM rat
        LEFT JOIN books AS b ON rat.book_id=b.book_id
        LEFT JOIN authors AS a ON b.author_id=a.author_id
        ORDER BY rat.mean_rat DESC
        LIMIT 5
        '''

pd.io.sql.read_sql(query, con = engine)

author  round
0         J.K. Rowling/Mary GrandPré   4.41
1                     J.R.R. Tolkien   4.39
2         J.K. Rowling/Mary GrandPré   4.29
3  Markus Zusak/Cao Xuân Việt Khương   4.26
4         J.K. Rowling/Mary GrandPré   4.25

#### Вывод по задаче

- самый **высокорейтинговый** писатель с оценкой **4,41** - **Джоан Роулинг**
- также указана в соавторстве **Мэри Грандпре** - иллюстратор книг Джоан Роулинг о Гарри Поттере
- значит культовые книги про мальчика, который выжил (Гарри Поттер) сделали эту пару самой рейтинговой

### Подсчет среднего количества обзоров от пользователей, которые поставили больше 50 оценок

In [12]:
query = '''
        WITH
        rev AS (SELECT username, 
                COUNT(*) AS count_reviews
        FROM reviews
        WHERE username IN (SELECT username
                           FROM ratings
                           GROUP BY username
                           HAVING COUNT(*) > 50)
        GROUP BY username)
        
        SELECT ROUND(AVG(count_reviews), 1)
        FROM rev

        '''
pd.io.sql.read_sql(query, con = engine)

round
0   24.3

#### Вывод по задаче

- в 24% случаев те, кто оценивает прочитанное оставляют и обзор на прочитанное. Почти каждый четвертый.